### ISSUE TITLE
[@typescript-eslint/no-unsafe-return] thinks value is any even when typed correctly

### ISSUE THREAD

**Issue by lifeiscontent** [typescript-eslint/typescript-eslint#3053]
- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-unsafe-return": "warning"
  }
}
```

```TS
import * as React from "react";

interface ChildProps {
  value: number;
}

const Child = (props: ChildProps) => null;
const Parent = ({
  children,
}: {
  children: React.ReactElement<ChildProps, () => null>[];
}) => {
  // options should be number[] but is any[] in ESLint running inside of VSCode
  const options = React.useMemo(
    () => React.Children.map(children, (child) => child.props.value) ?? [],
    [children]
  );

  return null;
};

const Example = () => (
  <Parent>
    <Child value={0} />
    <Child value={10} />
    <Child value={20} />
    <Child value={30} />
  </Parent>
);
```

**Expected Result**
options to be `number[]`

**Actual Result**
eslint says its being casted to `any[]`

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.15.0` |
| `@typescript-eslint/parser`        | `4.15.0` |
| `TypeScript`                       | `4.1.5` |
| `ESLint`                           | `7.19.0` |
| `node`                             | `14.15.5` |

**Comment by bradzacher** [comment]
I am unable to repro this.
Please provide a minimal repro.

![image](https://user-images.githubusercontent.com/7462525/107885752-da422700-6eb0-11eb-8db4-b5489c8166ff.png)

(the last example is a purposely failing case to prove the rule is working)

**Comment by lifeiscontent** [comment]
@bradzacher can I add you to my monorepo? eslint itself does not complain, its only when running VSCode which is super odd.

**Comment by lifeiscontent** [comment]
@bradzacher ping

**Comment by bradzacher** [comment]
I unfortunately as a volunteer maintainer, don't have the spare time to dig *that* deeply into this bug.
I barely have enough time to keep on top of issues and PRs here.

**Comment by lifeiscontent** [comment]
@bradzacher fair enough, thanks for the help anyway. I appreciate it!
